doc/developer-guides/GVT-g-kernel-options.rst

```diff
@@ -0,0 +1,237 @@
+.. _GVT-g-kernel-options:
+
+GVT-g (AcrnGT) Kernel Options
+#############################
+
+AcrnGT supports flexible configurations so customers can configure
+AcrnGT easily, according to their requirements. This document introduces
+the kernel command line options that are related to AcrnGT.
+
+i915.enable_gvt
+***************
+
+This option enables support for Intel GVT-g graphics virtualization
+support in the host By default, it's not enabled, so we need to add
+``i915.enable_gvt=1`` in the SOS kernel command line.  This is a Service
+OS only parameter, and cannot be enabled in the User OS.
+
+i915.enable_pvmmio
+******************
+
+We introduce the feature named **Para-Virtualized MMIO** (PVMMIO)
+to improve graphics performance of the GVT-g guest.
+This feature batches sequential MMIO writes into a
+shared buffer between the Service OS and User OS, and then submits a
+para-virtualized command to notify to GVT-g in Service OS. This
+effectively reduces the trap numbers of MMIO operations and improves
+overall graphics performance.
+
+The ``i915.enable_pvmmio`` option controls
+the optimization levels of the PVMMIO feature: each bit represents a
+sub-feature of the optimization. By default, all
+sub-features of PVMMIO are enabled but can also be selectively
+enabled as well.
+
+The PVMMIO optimization levels are:
+
+* PVMMIO_ELSP_SUBMIT = 0x1 - Batch submission of the guest graphics
+  workloads
+* PVMMIO_PLANE_UPDATE = 0x2 - Batch plane register update operations
+* PVMMIO_PLANE_WM_UPDATE = 0x4 - Batch watermark registers update operations
+* PVMMIO_MASTER_IRQ = 0x8 - Batch IRQ related registers
+* PVMMIO_PPGTT_UPDATE = 0x10 - Use PVMMIO method to update the PPGTT table
+  of guest.
+
+.. note:: This parameter works in both the Service OS and User OS, but
+   changes to one will affect the other. For example, if either SOS or UOS
+   disables the PVMMIO_PPGTT_UPDATE feature, this optimization will be
+   disabled for both.
+
+i915.gvt_workload_priority
+**************************
+
+acrnGT supports **Prioritized Rendering** as described in the
+:ref:`GVT-g-prioritized-rendering` high-level design.  This
+configuration option controls the priority level of GVT-g guests.
+Priority levels range from -1023 to 1023.
+
+The default priority is 0, the same priority as the Service OS. If the
+level is negative, the guest's priority will be lower than the Service
+OS, so graphics preemption will work and the prioritized rendering
+feature will be enabled.  This is a Service OS only parameters, and does
+not work in the User OS.
```

Comment:
Right now, different UOS share the same priority. We cannot assign different priorities to different UOS

Reply:
Thanks for the confirmation @minhe1 . The test indicates that if that number is negative, gfx preemption will work. I'm assuming it will work in the case the Gfx workload runs in the SOS, is that correct?

Does gfx preemption work from all UOS when you put a positive number there?

@dbkinder , it would be good to add some text based on those answers to make the description clearer and more complete.

Reply:
> The test indicates that if that number is negative, gfx preemption will work. I'm assuming it will work in the case the Gfx workload runs in the SOS, is that correct?
> Does gfx preemption work from all UOS when yo

If we set a negative value to this parameter, the SOS gfx workloads will preempt all the UOS gfx workloads.

Reply:
> If we set a negative value to this parameter, the SOS gfx workloads will preempt all the UOS gfx workloads.

Thanks for the confirmation. What is the effect of using a positive number (>0)?

Reply:
>             > If we set a negative value to this parameter, the SOS gfx workloads will preempt all the UOS gfx workloads.
> 
> 
> Thanks for the confirmation. What is the effect of using a positive number (>0)?
If set a positive number, the UOS gfx workloads will preempt most of the SOS gfx workloads, except the display updating related workloads, who have default highest priority 1023.